File: templates/skeleton/nestjs/src/example/example.controller.ts
import {
  Body,
  ConflictException,
  Controller,
  Get,
  NotFoundException,
  Param,
  Post,
} from '@nestjs/common';

import { CreateExampleDto } from './example.dto.js';
import type { Example } from './example.repository.js';
import { DuplicateExampleNameError, ExampleNotFoundError, ExampleService } from './example.service.js';

/**
 * HTTP layer of the example slice. Expected domain errors become status codes
 * here; the service stays framework-free (AGENTS.md A3).
 */
@Controller('examples')
export class ExampleController {
  constructor(private readonly service: ExampleService) {}

  /** Creates an example, rejecting a duplicate name with 409. */
  @Post()
  create(@Body() payload: CreateExampleDto): Example {
    try {
      return this.service.create(payload.name);
    } catch (error) {
      if (error instanceof DuplicateExampleNameError) {
        // Generic message to the client, detail stays in the log (A3, A5).
        throw new ConflictException('name already in use');
      }
      throw error;
    }
  }

  /** Returns one example, or 404 when it does not exist. */
  @Get(':id')
  read(@Param('id') id: string): Example {
    try {
      return this.service.get(id);
    } catch (error) {
      if (error instanceof ExampleNotFoundError) {
        throw new NotFoundException('not found');
      }
      throw error;
    }
  }
}

File: templates/skeleton/java-spring/src/main/java/com/example/@@PROJECT_SNAKE@@/api/ExampleController.java
package com.example.@@PROJECT_SNAKE@@.api;

import com.example.@@PROJECT_SNAKE@@.api.dto.CreateExampleRequest;
import com.example.@@PROJECT_SNAKE@@.api.dto.ExampleResponse;
import com.example.@@PROJECT_SNAKE@@.domain.ExampleService;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

/**
 * HTTP layer of the example slice.
 *
 * <p>The controller translates between HTTP and the service. Business rules
 * live in the service; expected domain errors become status codes in
 * {@link ApiExceptionHandler}.
 */
@RestController
@RequestMapping("/examples")
public class ExampleController {

    private final ExampleService service;

    /**
     * Creates the controller.
     *
     * @param service the example service
     */
    public ExampleController(ExampleService service) {
        this.service = service;
    }

    /**
     * Creates an example.
     *
     * @param request the validated payload
     * @return 201 with the created example
     */
    @PostMapping
    public ResponseEntity<ExampleResponse> create(@Valid @RequestBody CreateExampleRequest request) {
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ExampleResponse.from(service.create(request.name())));
    }

    /**
     * Returns one example.
     *
     * @param id the identifier to look up
     * @return 200 with the example
     */
    @GetMapping("/{id}")
    public ExampleResponse read(@PathVariable String id) {
        return ExampleResponse.from(service.get(id));
    }
}
